docs/src/adr/006rfc-unit-testing.md

```diff
@@ -0,0 +1,607 @@
+to isolate a set of executions to make continuous integration break, until the
+protocol is fixed. Needless to say, TLA+ tools have no support for this
+standard technique, though they have all capabilities to produce such tests.
+
+Similar to testing an action in isolation, we propose an interface for testing
+a restricted set of executions as follows:
+
+```tla
+{{#include ../../../test/tla/ChangRobertsTyped_Test.tla:116:127}}
+```
+
+In this case, we are using a different assertion in the `@ensure` annotation:
+
+```tla
+{{#include ../../../test/tla/ChangRobertsTyped_Test.tla:129:132}}
+```
+
+The test `TestExec_n0_n1` is similar to `TestAction_n0` in many aspects.  It
+starts by initializing the state with the predicate `Prepare_n0` and it expects
+a final state to satisfy the predicate `Assert_noWinner`. There is an important
+difference between the variables in `Assert_n0` and `Assert_noWinner`:
+
+ - Unprimed variables in `Assert_n0` refer to a state before firing an action,
+    whereas primed variables in `Assert_n0` refer to a state after firing
+    the action.
+
+ - Unprimed variables in `Assert_noWinner` refer to a state before firing
+   an *execution*, whereas primed variables in `Assert_noWinner` refer to
+   a final state of the execution.
+
+*(If you find the above behavior of `Assert_noWinner` confusing, please let us
+know.)*
+
+We should be able to run this test via:
+
+```sh
+apalache test --include=TestExec_n0_n1 ChangRobertsTyped_Test.tla
+```
+
+If the test is violated, a counterexample should be produced in the file
+`counterexample.tla`.
+
+### 3.4. Test executions with temporal properties
+
+When we wrote the test `TestExec_n0_n1`, we did not think about the
+intermediate states of an execution. This test was a functional test: It is
+matching the output against the input. When reasoning about state machines,
+we often like to restrict the executions and check the properties of those
+executions.
+
+Fortunately, we have all necessary ingredients in TLA+ to do
+exactly this. Test `TestExec_correctness_under_liveness`.
+
+```tla
+{{#include ../../../test/tla/ChangRobertsTyped_Test.tla:133:143}}
+```
+
+Predicates `Correctness` and `Liveness` are defined in the spec as follows:
+
+```tla
+{{#include ../../../test/tla/ChangRobertsTyped.tla:146:157}}
+```
+
+Since `Correctness` is a state predicate, we wrap it with a temporal operator
+to check it against all states of an execution:
+
+
+```tla
+{{#include ../../../test/tla/ChangRobertsTyped_Test.tla:144:145}}
+```
+
+
+
+### 3.5. Discussion
+
+As you can see, we clearly decompose a test in three parts:
+
+ - preparing the states (like a small version of `Init`),
+ - executing the action (like a small version of `Next`),
+ - testing the next states against the previous states (like an action invariant).
+
+In the rest of this section, we comment on the alternative approaches.
+
+#### 3.5.1. But I can do all of that in TLA+ 
```

Comment:
These sections make the particular feature set and approach very well motivated, IMO. Thanks for adding them!

Reply:
I was actually surprised how hard it was to express such a simple test in a dedicated operator, instead of creating a test in a separate file.